hw/ip_templates/ac_range_check/rtl/ac_range_check.sv.tpl

```diff
@@ -76,6 +79,198 @@ module ${module_instance_name}
+  // TLUL Loopback for failing accesses
+  //////////////////////////////////////////////////////////////////////////////
+
+  tlul_request_loopback u_req_loopback (
+    .clk_i         ( clk_i                 ),
+    .rst_ni        ( rst_ni                ),
+    .squash_req_i  ( range_check_fail      ),
+    // Incoming request
+    .tl_h2d_i      ( ctn_tl_h2d_i          ),
+    .tl_d2h_o      ( ctn_tl_d2h_o          ),
+    // Outgoing request/rsp if not squashed
+    .tl_h2d_o      ( ctn_filtered_tl_h2d_o ),
+    .tl_d2h_i      ( ctn_filtered_tl_d2h_i )
+  );
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Range Check Deny Counting Logic
+  //////////////////////////////////////////////////////////////////////////////
+
+  logic [DenyCountWidth-1:0] deny_cnt;
+  logic deny_cnt_incr, deny_cnt_clr;
+
+  // Only increment the deny counter if logging is enabled
+  assign deny_cnt_incr = reg2hw.log_config.log_enable.q & range_check_fail;
+  // Determine if we are doing the first log. This one is special, since it also needs to log
+  // diagnostics data
+  assign log_first_deny = deny_cnt_incr & (deny_cnt == 0);
+
+  // Clear log information when clearing the interrupt or when clearing the log manually via the
+  // the log_clear bit.
+  logic clear_log;
+  assign clear_log = (intr_state_q & ~reg2hw.intr_state.q) |
+                     (reg2hw.log_config.log_clear.qe & reg2hw.log_config.log_clear.q);
+
+  prim_count #(
+    .Width(DenyCountWidth)
+  ) u_deny_count (
+    .clk_i              ( clk_i              ),
+    .rst_ni             ( rst_ni             ),
+    .clr_i              ( clear_log          ),
+    .set_i              ( 1'b0               ),
+    .set_cnt_i          ( '0                 ),
+    .incr_en_i          ( deny_cnt_incr      ),
+    .decr_en_i          ( 1'b0               ),
+    .step_i             ( DenyCountWidth'(1) ),
+    .commit_i           ( 1'b1               ),
+    .cnt_o              ( deny_cnt           ),
+    .cnt_after_commit_o (                    ),
+    .err_o              ( deny_cnt_error     )
+  );
+
+  // Log count is transparently mirrored. Clearing happens on the counter.
+  assign hw2reg.log_status.deny_count.de = 1'b1;
+  assign hw2reg.log_status.deny_count.d  = deny_cnt;
+
+  assign hw2reg.log_status.denied_read_access.de = log_first_deny | clear_log;;
+  assign hw2reg.log_status.denied_read_access.d  = log_first_deny? read_access : 1'b0;
+
+  assign hw2reg.log_status.denied_write_access.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_write_access.d  = log_first_deny? write_access : 1'b0;
+
+  assign hw2reg.log_status.denied_execute_access.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_execute_access.d  = log_first_deny? execute_access : 1'b0;
+
+  assign hw2reg.log_status.denied_no_match.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_no_match.d  = 1'b0;
```

Comment:
I'm wondering if that makes even sense?

The log bits log a denied access that is either a read, wrote, or execute access - there are bit for these events. What does "denied-no-match" even mean?

Reply:
This would distinguish between (a) there was at least one matching range but it didn't allow the requested (read, write, or execute) access from (b) there was no range matching the address -- right?

Reply:
That at least, yes.